### DISCUSSION TITLE
How to create and run new instance? (Edit: simultaneously)

### DISCUSSION THREAD

**Discussion by SomwareHR** [TriliumNext/Trilium#6159] [Q&A]
Hi, first - thanks for this great software!

There are 9 issues that mention word "instanceName" but none of them is related to creating a new instance. Google was of no help.

"`config.ini`" file says..

```
 [General]
 # Instance name can be used to distinguish between different instances using backend api.getInstanceName()
 instanceName=

```
..so I conclude there is a possibility to run multiple Trilium instances on one server.

But - how?! :)

BR

**Comment by sigaloid** [comment]
You can also start a new instance by launching the server with the environment variable `TRILIUM_DATA_DIR=*dir*`.

**Comment by SomwareHR** [comment]
> You can also start a new instance by launching the server with the environment variable `TRILIUM_DATA_DIR=*dir*`.

Thanks. 
Your suggestion got me some ideas I'm exploring now.

1. export TRILIUM_DATA_DIR=/root/.local/share/trilium-data-second
2. "./node/bin/node src/www"
3. (it will fail - port 8080 is taken)
4. nano /root/.local/share/trilium-data-second/config.ini
5. change row "instanceName="
6. change row "port="
7. repeat "./node/bin/node src/www"

Seems it will work but throws an error

"Error when calling GET notes/%RANDOM%/backlink-count: error - Unauthorized"

and it can work either one or other, not both at the same time.
I could probably run it as another user?

Tnx
BR

**Comment by sigaloid** [comment]
Try setting the environment variable "TRILIUM_PORT" as well.

**Comment by SomwareHR** [comment]
_> Try setting the environment variable "TRILIUM_PORT" as well._

Thanks, it wont work that way :/

Secon Trilium runs after I edit "config.ini" on port 8081, and it listens to the old port 8080. But it can't serve both at the same time - it's either one or another. 

When you login to 8080 it works but error appears if you try to open 8081 ("Error when calling GET notes/%RANDOM%/backlink-count: error - Unauthorized"). 
Then you login to server on port 8081 and it works, but now the one on port 8080 shows same error.

So it's either first or second, but not both at the same time.
I'll see if it can work under another linux user account.

BR

**Comment by zadam** [comment]
Hi, this is because (authentication) cookies are shared between ports, so when you login to one instance on one port, these cookies will be used also for the other instance, but that's bound to fail. This is a general property of cookies/http.

To fix this you need to have trilium instances on either different (sub)domains or different paths, e.g. `domain.com/trilium1` and `domain.com/trilium2`.

**Comment by tritium21** [comment]
> To fix this you need to have trilium instances on either different (sub)domains or different paths, e.g. `domain.com/trilium1` and `domain.com/trilium2`.

It doesn't look like there is a way to tell trilium to use a different path for the cookie, unless i'm missing something.

**Comment by mzrimsek** [comment]
If you run them on different subdomains they should be two entirely unrelated instances. Assuming you have a subdomain pointed at your existing Trilium install and you are using Docker, all you should have to do is allocate a different subdomain for your 2nd instance (and point it at your new instance) and bind a different directory to your volume (or just let Docker create you a different volume I like binding folders to be able to easily back up files as such). If you are exposing a port for your trilium instance you'll need to bind an unallocated host port to your Docker container.

**Comment by tritium21** [comment]
> If you run them on different subdomains they should be two entirely unrelated instances. Assuming you have a subdomain pointed at your existing Trilium install and you are using Docker, all you should have to do is allocate a different subdomain for your 2nd instance (and point it at your new instance) and bind a different directory to your volume (or just let Docker create you a different volume I like binding folders to be able to easily back up files as such). If you are exposing a port for your trilium instance you'll need to bind an unallocated host port to your Docker container.

I think its understood that using a second domain is a solution to the problem, I think it is a non-viable solution for the original poster, and it is very much a non-viable solution for me.

**Comment by SomwareHR** [comment]
_> different subdomains_ 

Thanks, I'll try with running under different usernames and different ports, also start client under RunAs second username and / or different browser.
